ページング機能の作成

i18morino opened the pull request.

i18morino committed:
__readMap,__writeMap追加

i18morino committed:
プログラム修正、コメント追加

i18morino committed:
util.hmmに_readMap,_writeMap追加、コメントの追加

i18morino committed:
戻り値のある関数が戻り値のないものになっていたのを修正

i18morino committed:
memBase,memLen削除、pageTable追加

tctsigemura committed:
_mul32に致命的なバグ

	modified:   os/util/crt0.s

tctsigemura committed:
コメントの微修正

	modified:   os/fs/blkFile.cmm

tctsigemura committed:
TaC-CPU V3 で組込み用OSが動作する

	modified:   kernel/dispatcher.s
	modified:   kernel/kernel.cmm

i18morino committed:
mmInit編集

i18morino committed:
mm.cmmで')'が足りないとエラーが出てしまいます。

i18morino committed:
mmInit()編集、エラー修正

i18morino committed:
関数二つ作ったのですが詰まってしまったので助言お願いしたいです。

i18morino committed:
allocMem,freeMem編集、pmMainにビットマップの初期化を追加、その他ミスを修正

i18morino committed:
mm,pm編集,ページプール初期化ルーチン追加

i18morino committed:
ミス修正

i18morino committed:
pcbSizeとheapSizeをグローバル変数から削除

i18morino committed:
Update process.hmm

i18morino committed:
Update process.hmm

i18morino committed:
Update process.hmm

i18morino committed:
Update pm.cmm

i18morino committed:
sysExec改良

i18morino committed:
sysExec改良,その他修正

i18morino committed:
ページテーブルと参照しているフレーム解放の関数を追加

i18morino committed:
doc追加

i18morino committed:
peekB,W,pokeB,Wの作成

tctsigemura committed:
1. readFatがflushFatを呼出時にLBAの計算値が破壊されるバグを訂正
2. ルートディレクトリのセクタ数を定数ではなくBPBから計算するように変更

	modified:   os/fs/blkFile.cmm
	modified:   uSD/kernel.bin
	modified:   uSD/kernel0.bin

tctsigemura committed:
v3.3.5

	modified:   HISTORY.txt
	modified:   VERSION

i18morino committed:
peekB,W,pokeB,W,strCpyVp作成

i18morino committed:
コメント追加

i18morino committed:
Merge remote-tracking branch 'upstream/master'

i18morino committed:
Merge branch 'master' into i18morino

i18morino committed:
コメント追加,masterとmerge

i18morino committed:
allocFrameとする所をallocMemとしていた部分を修正

tctsigemura commented on os/mm/mm.cmm:
```diff
@@ -66,9 +70,17 @@ struct MemBlk {                                 // 空き領域管理用の構
+  //ページプールのアドレス
+  memPool.next.size = pagePoolAdr-_end();         //空きメモリサイズ
   memPool.next.next = null;
 }
```
定数はコンパイル時点で計算できるので変数ではなく```#define```の方が良いと思います．
```
#define PCBSIZ (sizeof(PCB)+P_KERN_STKSIZ+P_FILE_MAX*sizeof(int))
#define HEPSIZ (PRC_MAX*PCBSIZ+(FILEMAX*(sizeof(File)+BLKSIZ)))
```

tctsigemura commented on os/mm/mm.cmm:
```diff
@@ -66,9 +70,17 @@ struct MemBlk {                                 // 空き領域管理用の構
+  // ヒープ領域
+  pagePoolAdr = (_end()+heapSize+PAGESIZ-1) & ~(PAGESIZ-1);
+  //ページプールのアドレス
+  memPool.next.size = pagePoolAdr-_end();         //空きメモリサイズ
```
コメントが的を得ていないように思います．
```
   // ページプールのアドレス
   pagePoolAdr = (_end()+heapSize+PAGESIZ-1) & ~(PAGESIZ-1);
   //ヒープ領域のサイズ
   memPool.next.size = pagePoolAdr-_end(); 
```

tctsigemura commented on os/mm/mm.cmm:
```diff
@@ -144,6 +156,52 @@ int mmFree(void[] mem) {                        // 領域解放
+    }
+  }
+  return -1;                                        //　見つからなかったらエラー
+}
```
２重のif文を使うより条件を工夫した方が良いと思います．
```
public int alloc2Frame(){                           //　２フレーム連続してフレームを割り当て
   for(int num=0;num<256;num=num+1) {               //　頭から探索
     if(_readMap(bitmap, num) == 0 &&
        _readMap(bitmap, num+1) == 0) {             //　次のフレームが空いているか確認
       _writeMap(bitmap,num,1);                    
       _writeMap(bitmap,num+1,1);                    
       return num;                                  //　1フレーム目の番号を返す
     }
   }
   return -1;                                       //　見つからなかったらエラー
 }
```

tctsigemura commented on os/mm/mm.cmm:
```diff
@@ -49,6 +49,10 @@
 #include <syscall.hmm>
 #include "mm.hmm"
+#include <fsDef.hmm>
+#include <fs.hmm>
```
fs.hmmのインクルードは必要ないような気がします．

tctsigemura commented on os/mm/mm.cmm:
```diff
@@ -66,9 +70,17 @@ struct MemBlk {                                 // 空き領域管理用の構
 MemBlk memPool = {null, 0};                     // 空き領域リストの番兵
 public int _end();                              // カーネルの BBS 領域の最後
 
+int pagePoolAdr;                                // ページプールのアドレス
```
呼び名は「ページプール」ではなく「フレームプール」の方が適切な気がしてきました．

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -72,6 +72,9 @@
 #include "exeFile.hmm"
+#include <process.hmm>
+#include <file.hmm>
+#include <fsDef.hmm>
```
これら３つのincludeは不要では？

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -182,31 +185,98 @@ int readTextData() {
   return 0;                                         // 正常終了
 }
 
+//　仮装アドレスを物理アドレスに変換する
```
仮装行列の「仮装」になっています．

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -182,31 +185,98 @@ int readTextData() {
 }
 
+//　仮装アドレスを物理アドレスに変換する
+void[] _VtoP(void[] a, PCB pcb){                   
+  int[] pageTable = pcb.pageTable;;                 //　ページテーブルのアドレス
+  int data;                                         //　ページテーブルの中のデータ
+  int page = _AtoI(a) >> 8;                         //　アドレスaの上位8ビット取り出し
+  data = pageTable[page*2];                         //　ページテーブルの中身を取り出す
+  data = data >> 7;                                 //　最上位ビットを取り出す
+  if(data==1){
+    data = pageTable[page*2+1];                     //　フレーム番号を読み込み
+    data = data << 8;                               //　上位8ビットにフレーム番号を移動
+    int adr = _AtoI(a);                             
+    adr = adr & 0x00ff;                             //　アドレスの下8ビットをマスクする
+    data = data +adr;                               //  下8ビットは変化しない
+  }
+  return  _ItoA(data);                              //　アドレスに変換して返す
+}
```
こんな感じでどうでしょうか．
```
#define V(pcb, page) (pcb.pageTable[(page)*2+1]>>15)
#define FRAME(pcb, page) (pcb.pageTable[(page)*2+1]&0xff)

void[] _VtoP(void[] a, PCB pcb){                   
   int page = _AtoI(a) >> 8;                         //　アドレスaの上位8ビット取り出し
   int adr  = _AtoI(a) & 0xff;                       //　アドレスaの下位8ビット取り出し
   if(V(pcb, page)==0) return null;                  // 変換できなかった
   return  _ItoA((FRAME(pcb, page)<<8)|adr);         //　アドレスに変換して返す
 }
```

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -182,31 +185,98 @@ int readTextData() {
+//　指定したアドレスの1バイトを返す
+char peekB(void[] a, PCB pcb){
+  char[] adr = _VtoP(a,pcb);                        //　指定したアドレスを物理アドレスに変換
+  if(_AtoI(adr)==0){                                //　フレームに変換できなかったら
+    exit(1);                                        //　エラー処理
+  }
```
```
  if(adr==null) exit(1);                             //　フレームに変換できなかった
```

tctsigemura replied:
pmを終了することになるので```exit(1);```してはいけない！！！
pmではなくpcbのプロセスを終了しなければならない．

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -182,31 +185,98 @@ int readTextData() {
+//　指定したアドレスの1ワードを返す
+int peekW(void[] a, PCB pcb){
+  char[] adr = _VtoP(a,pcb);                        //　指定したアドレスを物理アドレスに変換
+  if(_AtoI(adr)==0){                                //　フレームに変換できなかったら
+    exit(1);                                        //  エラー処理
+  }
```
208行と同様

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -182,31 +185,98 @@ int readTextData() {
+
+//　指定したアドレスの1ワードを返す
+int peekW(void[] a, PCB pcb){
+  char[] adr = _VtoP(a,pcb);                        //　指定したアドレスを物理アドレスに変換
```
ardの型を```int[]```にしておけば，220行から先が簡単になる．

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -182,31 +185,98 @@ int readTextData() {
+  if(_AtoI(adr)==0){                                //　フレームに変換できなかったら
+    exit(1);                                        //  エラー処理
+  }
+  char word1 = adr[0];                              //  アドレスの中身1バイト
+  char word2 = adr[1];                              //　アドレスの一つ次の中身1バイト
+  return (ord(word1) << 8 | ord(word2));            //　1ワード(2バイト)にして返す
```
adrの型が```int[]```なら次の１行で良い．
```return adr[0];```

tctsigemura reviewed (commented):
見たところまでで一度サブミットします．

i18morino committed:
現時点でのプログラムです

tctsigemura commented on os/mm/mm.cmm:
```diff
@@ -66,21 +65,22 @@ struct MemBlk {                                 // 空き領域管理用の構
 
-int pagePoolAdr;                                // ページプールのアドレス
+int framePoolAdr;                                // フレームプールのアドレス
 
```
コメントのインデントがずれました．

tctsigemura commented on os/mm/mm.cmm:
```diff
@@ -195,7 +195,7 @@ public void freeFrame(int num){                     //　指定したフレー
+  int kPage = framePoolAdr/PAGESIZ;  //カーネルのページ数
   for(int cnt=0;cnt<kPage;cnt=cnt+1){               //ページ数分フレームを割り当てる
     allocFrame();
   }
```
```allocFrame();```だと
「allocFrameが先頭から順にフレームを使用する」という知識が無いと意味がわかりにくので，
```_writeMap(bitmap, cnt, 1);```
の方が良いように思います．

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -185,57 +183,59 @@ int readTextData() {
+    pcb.pageTable[page*2+1] = num;                  //　フレーム番号をページテーブルに記憶
   }
   int word1 = i >> 8;                               //　前1バイトを取り出し
   int word2 = i & 0x00ff;                           //　後ろ１バイトを取り出し
```
メモリが厳しくてプログラムをなるべくコンパクトにしたいので，
```pokeB()```を2回呼び出すようにしましょう．

tctsigemura replied:
いやいや，アドレス変換部分をサブルーチン化して
pokeBとpokeWの両方で使用するようにするが正解な気がしてきました．

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -252,11 +252,22 @@ void strCpyVP(void[] dst, void[] src, PCB pPcb, PCB cPcb){
   }
 }
 
+//　スワップイン
+void swapIn(void[] adr, PCB pcb, int fnum){
+  int page = _AtoI(adr)>>8;                         //　スワップインするアドレスのページ番号
+  int[] fad = _ItoA(fnum<<8);                       //　フレーム番号からアドレスを求める
+  fd = pcb.filedsc;                                 //　pcbのexeファイルのファイルディスクリプタ
+
+  for(int num=0;num<256;num=num+1){
+    fad[num] = getByte();
+  }
+}
```
260行あたりでexeファイルを目的のページ位置にseekする必要があります．

tctsigemura replied:
getByteを繰り返し使用しなくても
```read(fd, fda, PAGESIZ);```
で良いような気がします．

tctsigemura replied:
```fd = ...```は```int fd = ...```かな？

tctsigemura replied:
readのエラーチェックをした方が良さそうです．
とりあえずはpanicにするだけでよいですが．

tctsigemura replied:
PAGESIZ 単位での読み出ししか必要なくなったので，
buf, openByte, getByte, closeByte は使用しなくなったのでは？

tctsigemura replied:
ヘッダを読むときにワード単位が残っていました．
ここは,```read(fd, buf, 2)```で良いと思います．
bufは```int[] buf = array(1);```で確保する．

tctsigemura replied:
seekの使用方法
1. fs.hmm をインクルードする．
2. 次のプログラムでできる．(posはファイル内のバイト位置です．16bitで表現できないときは第２引数が0ではなくなりますが，.exeファイルが64KiBより大きくなることはないので第２引数はいつも0です．）
```seek(fd, 0, pos);```

tctsigemura reviewed (commented):
みました．

i18morino committed:
細かな修正,openByte等コメントアウト

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -141,16 +141,17 @@ int bssSiz;                                         // BSS  セグメントサ
+  magic   = read(fd, buf, 2);                       // マジック番号
   if (magic!=UMAGIC && magic!=IMAGIC)               // マジック番号をチェック
     return EMAGIC;                                  //   不正なマジック番号
-  textSiz     = getWord();                          // TEXT セグメントサイズ
-  dataSiz     = getWord();                          // DATA セグメントサイズ
-  bssSiz      = getWord();                          // BSS  セグメントサイズ
-  relSiz      = getWord();                          // 再配置情報のサイズ
-  stkSiz      = getWord();                          // ユーザスタックサイズ
+  textSiz     = read(fd, buf, 2);                   // TEXT セグメントサイズ
+  dataSiz     = read(fd, buf, 2);                   // DATA セグメントサイズ
+  bssSiz      = read(fd, buf, 2);                   // BSS  セグメントサイズ
+  relSiz      = read(fd, buf, 2);                   // 再配置情報のサイズ
+  stkSiz      = read(fd, buf, 2);                   // ユーザスタックサイズ
```
```
magic   = read(fd, buf, 2);
```
ではなく，
```
read(fd, buf, 2);
magic   = buf[0];
```
となります．
getWord関数を次の様に復活すると以前のように```magic = getWord();```で済みます．
```
int getWord() {
  read(fd, buf, 2);
  return buf[0];
}
```

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -212,35 +213,33 @@ int peekW(void[] a, PCB pcb){
-  adr[1] = word2;                                   //　次のアドレスに1ワードの後ろ1バイトを格納
+  adr[0] = word >> 8;                               //  アドレスに1ワードの前１バイトを格納
+  adr[1] = word & 0x00ff;                           //　次のアドレスに1ワードの後ろ1バイトを格納
 }
```
241,242行の adr は ```int[]```型だから，```adr[0] = word;```でよいです．

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -254,20 +253,21 @@ void strCpyVP(void[] dst, void[] src, PCB pPcb, PCB cPcb){
-  }
+  int[] fadr = _ItoA(fnum<<8);                      //　フレーム番号からアドレスを求める
+  int fd = pcb.filedsc;                             //　pcbのexeファイルのファイルディスクリプタ
+  r = seek(fd, 0, _AtoI(adr));                      // 入力を指定のアドレスまでSeek
+  if(r<0) panic("result error");
+  r = read(fd, fadr, PAGESIZ);                       //　ページ単位のコピー
+  if(r<0) panic("result error");
```
panicのメッセージは```seek error```，```read error```にした方がデバッグの役に立つと思います．

tctsigemura replied:
1. seekする位置はヘッダサイズ分，後ろにずらす必要があると思います．
2. adrはページサイズの整数倍なのか？　まだ，決まっていないので，```_AtoI(adr)```ではなく```page*PAGESIZ```の方が良いと思います．

tctsigemura replied:
それからアドレスとページ番号の変換は８ビットシフトではなく，
PAGESIZを掛けたり割ったりして求めてください．
将来PAGESIZを512に変更したりしたときのためになります．
また，コンパイラが2のべき乗の乗除はシフトに置き換えるので
出来上がる機械語はシフトも２のべき乗の乗除も同じです．

tctsigemura reviewed (commented):
コメントを確認してください．

i18morino committed:
コメントの内容を修正

tctsigemura committed:
組み込みカーネルで不要な部分を#ifdefで囲む

	modified:   kernel/kernel.cmm
	modified:   kernel/process.hmm

tctsigemura committed:
addrof(_end) と end() との書き間違いを訂正

	modified:   mm/mm.cmm

tctsigemura committed:
組み込みカーネルが動作するように暫定的な変更
！！！PCBを変更した場合は dispatcher.s にも変更が必要！！！

	modified:   os/kernel/dispatcher.s
	modified:   os/kernel/process.hmm

tctsigemura committed:
Merge branch 'master' into i18morino

i18morino committed:
行き詰まってしまいました。疑問に思ったことをプログラムにコメントで書いています。

i18morino committed:
Merge branch 'i18morino' of https://github.com/TecPaging/TacOS into i18morino

i18morino committed:
混乱を整理

i18morino committed:
コメントアウトした部分の削除と一部修正

i18morino committed:
引数の型を変更

i18morino committed:
_AtoP,_ItoP削除のため_VtoPへ変更、ページテーブル操作の修正

tctsigemura commented on os/mm/mm.cmm:
```diff
@@ -49,6 +49,9 @@
 #include <syscall.hmm>
 #include "mm.hmm"
+#include <fsDef.hmm>
+#include <process.hmm>
```
process.hmm のインクルードは不要かも

tctsigemura committed:
ソースコードの体裁を整えた

	modified:   mm/mm.cmm
	modified:   mm/mm.hmm
	modified:   pm/pm.cmm
	modified:   util/util.hmm

i18morino committed:
インデントの修正

i18morino committed:
frameChange作成

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -148,65 +112,156 @@ int readHeader() {
+    if (flag == 1) break;
   }
-  return 0;                                         // 正常終了
 }
```
1. 2重ループは止めて，各プロセスについて「解放できるフレームを探す関数」を呼び出すようにしましょう．
2. procTblの中にはnullになっているものもあるはずです．
3. 「プロセスの全てのページに付いて」はPAGESIZ回の繰り返しではないはずです．
4. 「プロセスの全てのページに付いて」ではなくTEXTセグメントの全てのページに付いてにすべきです．

tctsigemura reviewed (commented):
frameChangeを見直してください．

i18morino committed:
frameChange編集,searchFrame作成

i18morino committed:
細かな修正

i18morino committed:
swapInを使用,関数の位置を変更

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -212,26 +212,34 @@ void swapIn(int adr, PCB pcb, int fnum){
+      } 
+    }
+  return fNum;
+ }
```
1. インデントが変です．　
216行は1文字深すぎます．
218行以下は217行と同じインデントです．
2. pcb.textSizはページ番号で管理していましたか？
アドレスではありませんか？

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -148,65 +112,168 @@ int readHeader() {
+  pEnd  = pBss + bssSiz;                            // 空間最後のアドレス
+}
+
+#define V(pcb, page) (pcb.pageTable[page]>>15)      // ページテーブルのVビット
```
Vは論理型にした方が良いと思います．
```
#define V(pcb, page) ((pcb.pageTable[page]&0x8000)!=0)
```

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -148,65 +112,168 @@ int readHeader() {
+int _VtoP(int a, PCB pcb){                   
+  int page = a / PAGESIZ;                           // ページ番号
+  int adr  = a % PAGESIZ;                           // ページ内アドレス
+  if (V(pcb, page)==0) return 0;                    // 変換できなかった
```
Vが論理型なら
```
if (!(V(pcb, page)) return 0;
```

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -148,65 +112,168 @@ int readHeader() {
+    // Textセグメントの全てのページについて
+    for(int page=0; page<pcb.textSiz; page=page+1) {
+      // V=1のフレームを入れ替える
+      if (V(pcb,page)==1) {
```
Vが論理型なら
```
if (V(pcb, page)) {
```

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -224,65 +291,91 @@ void copyArgsEnvp(char[][] argv, char[][] envp, PCB pcb) {
+  int entry;
+  // フレームの解放
+  for(int num; num<256; num=num+1) {                // ページテーブルのidx
+    entry = V(pcb, num);                            // エントリのVビットを確認
+    if(entry==1){                                   // Vビットが1なら
+      entry = pageTable[num]&0x00ff;                // フレーム番号取り出し
+      freeFrame(entry);                             // フレーム解放
+    }
```
Vが論理型かつentryのスコープを小さくすると
```
    if (V(pcb, num)) {
       int entry = pageTable[num]&0x00ff;
       freeFrame(entry);
     }
```
または，
```
    if (V(pcb, num)) {
       freeFrame(pageTable[num]&0x00ff);
     }
```

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -224,65 +291,91 @@ void copyArgsEnvp(char[][] argv, char[][] envp, PCB pcb) {
+      freeFrame(entry);                             // フレーム解放
+    }
   }
+  // ページテーブルの解放
+  int pageFrame = _AtoI(pcb.pageTable);
+  pageFrame = pageFrame/PAGESIZ;
+  freeFrame(pageFrame);
```
1. ページテーブルには２フレーム使用しているのでは？
2. ```pcb.pageTable```のコピーが```pageTable```変数にあるのでそっちを使うべき．
3. 256はPAGESIZと同時にTOTALPAGEのような記号定数にしておく．

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -224,65 +291,91 @@ void copyArgsEnvp(char[][] argv, char[][] envp, PCB pcb) {
+   int pageTableAdr = pageFrame * PAGESIZ;          // フレーム番号をアドレスに
+   newPcb.pageTable = _ItoA(pageTableAdr);          // PCBにページテーブル登録
+   int[] pageTable = _ItoA(pageTableAdr);
+   for(int num=0; num<256; num=num+1){              // ページテーブルを初期化
```
ここも256は記号定数にする．

tctsigemura reviewed (commented):
見たところまでで，もう一度，送ります．

i18morino committed:
軽微な修正、コメント追加

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -132,37 +132,38 @@ void locateSeg() {
+  }
   return fNum;                                      // フレーム番号を返す
- }
+}
```
見つからなかった場合はゼロを返す仕様ですね．
コメントに一言書いておくと良いと思います．

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -132,37 +132,38 @@ void locateSeg() {
+}
 
 //  フレームの空きが無い時にフレームを入れ替える
 int frameChange(PCB pcb, int adr){
```
```
int frameChange(PCB pcb, int adr){
  PCB p;
  int fNum = 0;
  int pIdx = adr/PAGESIZ;                           // adrのページテーブルのidx
  PCB p;                                            // プロセスのPCB
  int fNum = 0;                                     // フレーム番号
  for(int i=0; i<PRC_MAX; i=i+1) {                  // 全てのプロセスについて
    p = procTbl[i];                                 // PCBを取り出し
    if(p != null) {                                 // PCBが入っていれば
      fNum = searchFrame(pcb);                      // 解放できるフレームを探す 
      if(fNum != 0) break;                          // フレームが見つかったらbreak
    }
  }
  return fNum;                                      // フレーム番号を返す
}
```
1. 変数```p```はfor文内部のローカル変数にしましょう．
2. ```searchFrame(pcb)```は```searchFrame(p)```の間違い？
3. 引数```pcb```,```adr```と変数```pldx```は不要？

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -332,7 +330,7 @@ int sysExec(int path, int argv, int envp, PCB pcb) {
    int pageTableAdr = pageFrame * PAGESIZ;          // フレーム番号をアドレスに
    newPcb.pageTable = _ItoA(pageTableAdr);          // PCBにページテーブル登録
    int[] pageTable = _ItoA(pageTableAdr);
```
331行と332行の順番を逆にすれば，```_ItoA```の呼び出しを減らすことができます．
```
newPcb.pageTable = pageTable;          // PCBにページテーブル登録
```

tctsigemura reviewed (commented):
一旦ここまでで．

tctsigemura commented:
sysExecの中で早めにPCBをmallocしています．
その後，ファイルのオープンに失敗したときやヘッダを読めなかった時，
PCBをfreeして終了する必要があります．
PCBのmallocをreadHeaderの後にした方が良いように思います．
その際，newProcとの役割分担を見直した方が良さそうです．

i18morino committed:
TLB Missの処理の実装,utilにルーチン追加

i18morino committed:
swapIn関連の変更

i18morino committed:
コメントの編集

i18morino committed:
swapInの変更,pokeWシステムコール作成

tctsigemura committed:
ファイル名を受け取るシステムコールの書き方の例

	modified:   os/fs/fs.cmm

i18morino committed:
peekS削除,Sビット操作,swapinシステムコール作成

i18morino committed:
peekS復元,pokeM実装

i18morino committed:
_ItoP,_AtoP削除

tctsigemura commented on os/pm/pm.cmm:
```diff
@@ -397,6 +400,31 @@ int sysExec(int path, int argv, int envp, PCB pcb) {
+    userAdr[i] = pAdr[i];
+  }
+  return 0;
+}
```
pokeMのuAdrはプロセスのユーザ空間のアドレスだから，
pokeBかpokeWを使って書き込む必要があるのでは？

tctsigemura replied:
その前に，pokeMはpmのコンテキストで動かなければならないのでシステムコールにするべき．．．
peekSも同様にシステムコールにするべきでは？
これらは，カーネルモードに切り替わったあとでユーザプロセスが呼び出す隠れシステムコール

tctsigemura reviewed (commented):
紙にまとめたものを研究室の机の上に置いておきます．

tctsigemura commented:
紙にまとめると書きましたが，以下にまとめます．
pokeM，peakSに関する変更点．
pokeM, peakS はPMサーバのコンテキストで実行されるべきなので，
1. ```pokeM```関数の名前は```sysPokeM```に```peekS```関数の名前は```sysPeekS```に変更する．
2. pmSysCallが```sysPokeM```，```sysPeekS```を呼び出すようにする．
3. execルーチンの直前あたりにユーザプロセスのコンテキストで呼び出される```pokeM```，```peekS```関数を追加し，sendrecを使用してPMサーバのコンテキストで```sysPokeM```，```sysPeekS```を実行させる．

tctsigemura commented:
pokeB, pokeWの微修正
pokeB, pokeWはフレームを確保するためにswapInを呼び出すようにする．
理由
1. プログラムが短くなる．
2. 将来，データ部分もswap-outの対象になった時，swapInを使うだろう．

ただし，vAdrがpcb.txtSize未満のときは書き込み禁止なのでpanicにする．

tctsigemura commented:
swapInについて
SWAPINはLOADに名前を変更する
1. LOADのときはsysLoadを呼び出すようにする．
2. sysLoad関数を新設し，この関数内でswapInを呼び出した上でページテーブのstickyをONにする．
（stickyを同時にONにしないとpmからユーザプロセスに切り替わる間にもswap-outされる可能性がある．）
3. execの前辺りにload関数を追加し，ここからsendrecでLOADシステムコールを呼び出す．
4. pm以外からSWAPINを利用していた部分はload関数の呼び出しに変更する．
5. 従来，pm内部でswapInを呼び出していた部分はそのままにしておく．

tctsigemura commented:
RELEASEについて（stickyのビットをOFFにする関数）

pmのexecの前辺りにrelease関数を追加する．
これはユーザプロセスのコンテキストで全て処理できる．
curProcのページテーブルの該当エントリのstickyビットをOFFにする．

i18morino committed:
peekS,pokeMシステムコール作成、read,write編集

i18morino committed:
loadシステムコール使用

i18morino committed:
release関数の引数変更,TLBミス部分の修正

i18morino committed:
readWrite作成

i18morino committed:
ttyRead,ttyWrite編集、peekMシステムコール作成

i18morino committed:
システムコール変更に伴うコメントの変更

tctsigemura committed:
コメントに関数名の綴間違い(_uComp(誤) → _uCmp(正))

	modified:   os/util/util.hmm

tctsigemura committed:
.bin を追加

	modified:   .gitignore

i18morino committed:
数値だったものを定数に変更

tctsigemura committed:
Merge branch 'i18morino' of https://github.com/TecPaging/TacOS into i18morino

tctsigemura committed:
カール内からのシステムコール発行を考慮

	modified:   os/fs/fs.cmm

tctsigemura committed:
サーバプロセスのPIDの最大値を記録する

	modified:   os/kernel/kernel.cmm
	modified:   os/kernel/process.hmm

tctsigemura committed:
peekM, pokeM の pAdrの型を void[] に変更
不要なグローバル変数を削除

	modified:   os/pm/pm.cmm
	modified:   os/pm/pm.hmm

tctsigemura committed:
Merge branch 'sigeExam' into i18morino

	modified:   os/fs/fs.cmm
	modified:   os/kernel/kernel.cmm
	modified:   os/kernel/process.hmm
	modified:   os/pm/pm.cmm
	modified:   os/pm/pm.hmm

tctsigemura committed:
マージ時のミス訂正

	modified:   os/pm/pm.cmm

tctsigemura committed:
コードレビュー
changeFrame を searchAllFrameに名称変更
swapInをlocateFrameと名称変更
swapInはファイルから読み込む機能だけの関数に変更
locateFrameはスワップインをする必要があるときだけswapInを呼び出す．

	modified:   os/pm/pm.cmm

i18morino committed:
copyArgsEnvpの現状

tctsigemura committed:
argv, envp を親プロセスから子プロセスのスタックにコピーする

	modified:   kernel/kernel.hmm
	modified:   pm/exeFile.hmm
	modified:   pm/pm.cmm

tctsigemura committed:
Revert "copyArgsEnvpの現状"

This reverts commit 6581e608cd8965bb988fbc39536ffca9faabdce4.

	modified:   os/pm/pm.cmm

tctsigemura committed:
Merge branch 'sigeWork' into i18morino(argv, envpを子プロセスにコピー）

tctsigemura committed:
ページング仮想記憶の機能を一通りコーディングした

	modified:   os/fs/fs.cmm
	modified:   os/kernel/dispatcher.s
	modified:   os/kernel/kernel.cmm
	modified:   os/mm/mm.cmm
	modified:   os/mm/mm.hmm
	modified:   os/pm/pm.cmm
	modified:   os/pm/pm.hmm

tctsigemura committed:
ページング機能の仕上げをマージ i18morino

tctsigemura committed:
仮想空間をアクセスするユーザプロセス用のシステムコールを分離

	modified:   os/GNUmakefile
	modified:   os/fs/fs.cmm
	modified:   os/kernel/init.cmm
	modified:   os/kernel/kernel.cmm
	modified:   os/kernel/process.hmm
	modified:   os/kernel/trap.s
	modified:   os/mm/mm.cmm
	modified:   os/pm/pm.cmm
	modified:   os/tty/tty.cmm
	modified:   os/tty/tty.hmm
	modified:   usr/sedit/GNUmakefile
	modified:   usr/tasm7/Makefile

tctsigemura committed:
仮想空間をアクセスするシステムコールに関する変更をマージ

tctsigemura committed:
フレームプール管理をmmからpmに移動，アドレスの割り算や大小比較で符号なしを徹底

	modified:   os/kernel/kernel.cmm
	modified:   os/mm/mm.cmm
	modified:   os/mm/mm.hmm
	modified:   os/pm/pm.cmm
	modified:   os/pm/pm.hmm
	modified:   os/util/crt0.s
	modified:   os/util/util.hmm

i18morino committed:
adrToOffsを使用するように変更

tctsigemura committed:
Usageメッセージにargv[0]を渡していなかった

	modified:   usr/rm/rm.cmm

tctsigemura committed:
アドレスをintに変換して計算する際，符号付きで計算される問題を解決
デバッグ表示を強化

	modified:   os/GNUmakefile
	modified:   os/fs/fs.cmm
	modified:   os/kernel/dispatcher.s
	modified:   os/kernel/kernel.cmm
	modified:   os/kernel/process.hmm
	modified:   os/kernel/syscall.hmm
	modified:   os/kernel/trap.s
	modified:   os/pm/pm.cmm
	modified:   os/sio/sio.cmm
	modified:   os/util/util.cmm
	modified:   os/util/util.hmm

i18morino committed:
rmFlagの条件を変更

i18morino committed:
sigFsmのバグを修正

tctsigemura committed:
アドレスの比較は符号なしにする必要がある

	modified:   ../os/kernel/trap.s
	modified:   ../os/pm/pm.cmm

tctsigemura committed:
Merge branch 'sigeWork' into i18morino

i18morino committed:
ページ入れ替えのバグを修正

i18morino committed:
デバッグの時に書き換えたけど戻すのを忘れたものを直した

i18morino committed:
0x38に現在のpidを書き込むように変更

i18morino committed:
シミュレータのためのpid書き込み+必要のないTLBflushをしないように変更

i18morino committed:
カーネルスタックサイズ変更+フレーム解放部分の改良

i18morino committed:
クラス図編集

i18morino committed:
保護ビットをTEXTとTEXT以外で変わるように変更

i18morino committed:
バグを修正

i18morino committed:
コメントの追加

i18morino committed:
発表，卒論で使用したdrawioのファイル

tctsigemura committed:
VERSIONコマンド追加

	new file:   uSD/bin/version.exe
	new file:   usr/version/GNUmakefile
	new file:   usr/version/Makefile
	new file:   usr/version/version.cmm

tctsigemura committed:
ページング方式仮想記憶が使用できる

	modified:   HISTORY.txt
	modified:   VERSION
	modified:   uSD/bin/Tasm7.exe
	modified:   uSD/bin/ar.exe
	modified:   uSD/bin/cat.exe
	modified:   uSD/bin/cp.exe
	modified:   uSD/bin/dump.exe
	modified:   uSD/bin/echo.exe
	modified:   uSD/bin/edit.exe
	modified:   uSD/bin/env.exe
	modified:   uSD/bin/fsize.exe
	modified:   uSD/bin/head.exe
	modified:   uSD/bin/hello.exe
	modified:   uSD/bin/in.exe
	modified:   uSD/bin/ls.exe
	modified:   uSD/bin/mkdir.exe
	modified:   uSD/bin/mv.exe
	modified:   uSD/bin/out.exe
	modified:   uSD/bin/printenv.exe
	modified:   uSD/bin/pwd.exe
	modified:   uSD/bin/rm.exe
	modified:   uSD/bin/rmdir.exe
	modified:   uSD/bin/screen.exe
	modified:   uSD/bin/sedit.exe
	modified:   uSD/bin/shell.exe
	modified:   uSD/bin/spitest.exe
	modified:   uSD/bin/sr.exe
	modified:   uSD/bin/tail.exe
	modified:   uSD/bin/touch.exe
	modified:   uSD/kernel.bin
	modified:   uSD/kernel0.bin
	modified:   usr/GNUmakefile
